File: src/kohakulayout/ir/text/writer.py
"""Levels to canonical text: sorted ids, fixed option order, moves for wires when the problem is known."""

from fractions import Fraction
from typing import Any

from kohakulayout._rust_bridge import rust_write_kl
from kohakulayout.ir.assessment import Assessment
from kohakulayout.ir.base import Level
from kohakulayout.ir.fabric import Fabric, Region
from kohakulayout.ir.layout import Layout, Wire
from kohakulayout.ir.netlist import Cell, Macro, Module, Net, Netlist, Pin, PinRef
from kohakulayout.ir.problem import Problem
from kohakulayout.ir.text.moves import moves_from_cells
from kohakulayout.ir.text.values import quote, write_loose_value, write_value

INDENT = "    "


def _opt(key: str, value: Any) -> str:
    return f"{key}={write_value(value)}"


def _attrs(attrs: dict[str, dict[str, Any]]) -> list[str]:
    return [
        f"+{ns}.{key}={write_loose_value(value)}"
        for ns in sorted(attrs)
        for key, value in sorted(attrs[ns].items())
    ]


def _rate(value: Fraction) -> str:
    return (
        str(value.numerator)
        if value.denominator == 1
        else f"{value.numerator}/{value.denominator}"
    )


# ------------------------------------------------------------------ fabric


def _region_line(region: Region, fabric: Fabric) -> str:
    every = fabric.width * fabric.height
    if len(region.cells()) == every:
        body = "all"
    else:
        body = " ".join(f"rect {r.x},{r.y} {r.w}x{r.h}" for r in region.rects)
    return " ".join(["region", region.id, body, *_attrs(region.attrs)])


def fabric_lines(fabric: Fabric) -> list[str]:
    parts = [
        "fabric",
        f"{fabric.width}x{fabric.height}",
        _opt("layers", list(fabric.layers)),
    ]
    if fabric.entries:
        parts.append(_opt("entries", list(fabric.entries)))
    lines = [" ".join([*parts, *_attrs(fabric.attrs)])]
    for rid in sorted(fabric.regions):
        lines.append(_region_line(fabric.regions[rid], fabric))
    for cid in sorted(fabric.carriers):
        carrier = fabric.carriers[cid]
        parts = ["carrier", carrier.id, carrier.layer]
        if carrier.capacity is not None:
            parts.append(_rate(carrier.capacity))
        lines.append(" ".join([*parts, *_attrs(carrier.attrs)]))
    return lines


# ----------------------------------------------------------------- netlist


def _port_line(port: Any) -> str:
    head = f"{port.id} {port.direction} {port.carrier} {port.side}{port.offset}"
    return " ".join([head, *_attrs(port.attrs)])


def _lib_lines(netlist: Netlist, indent: str) -> list[str]:
    lines = []
    for fid in sorted(netlist.library):
        fp = netlist.library[fid]
        parts = ["lib", fp.id, f"{fp.width}x{fp.height}"]
        if fp.layer != "ground":
            parts.append(_opt("layer", fp.layer))
        if fp.occludes:
            parts.append(_opt("occludes", list(fp.occludes)))
        if tuple(fp.rotations) != (0, 90, 180, 270):
            parts.append(_opt("rot", list(fp.rotations)))
        parts += _attrs(fp.attrs)
        ports = [_port_line(p) for p in fp.ports]
        if len(ports) <= 4:
            lines.append(
                f"{indent}{' '.join(parts)} {{ {' ; '.join(ports)} }}"
                if ports
                else f"{indent}{' '.join(parts)} {{ }}"
            )
        else:
            lines.append(f"{indent}{' '.join(parts)} {{")
            lines += [f"{indent}{INDENT}{p}" for p in ports]
            lines.append(f"{indent}}}")
    return lines


def _cell_line(cell: Cell, pack: str) -> str:
    ref = cell.footprint or cell.module or cell.macro or ""
    parts = ["cell", cell.id, ref]
    if cell.kind and cell.kind != ref:
        parts.append(_opt("kind", cell.kind))
    if cell.label:
        parts.append(_opt("label", cell.label))
    constraint = cell.constraint
    if constraint.kind != "free" or constraint.attrs:
        parts.append(_opt("at", constraint.kind))
        for ns in sorted(constraint.attrs):
            for key, value in sorted(constraint.attrs[ns].items()):
                parts.append(
                    _opt(key if ns == (pack or "pack") else f"at.{ns}.{key}", value)
                )
    if cell.group:
        parts.append(_opt("group", cell.group))
    if cell.needs:
        parts.append(_opt("needs", list(cell.needs)))
    parts += _attrs(cell.attrs)
    return " ".join(parts)


def _default_pins(netlist: Netlist, cell: Cell) -> tuple[Pin, ...]:
    return netlist.model_copy(
        update={
            "cells": {**netlist.cells, cell.id: cell.model_copy(update={"pins": ()})}
        }
    ).pins_of(cell.id)


def _pin_lines(netlist: Netlist, cell: Cell) -> list[str]:
    if not cell.pins or cell.pins == _default_pins(netlist, cell):
        return []
    return [
        f"pin {cell.id}.{p.id} {p.direction} {p.carrier} {_opt('ports', '|'.join(p.ports))}"
        for p in cell.pins
    ]


def _net_line(net: Net) -> str:
    parts = ["net", net.id, net.carrier]
    if net.rate:
        parts.append(_rate(net.rate))
    if net.kind:
        parts.append(_opt("kind", net.kind))
    if net.label:
        parts.append(_opt("label", net.label))
    if net.outside is not None:
        parts.append(_opt("outside", net.outside))
    parts += _attrs(net.attrs)
    parts.append(":")
    parts += [str(r) for r in net.sources]
    parts.append("->")
    parts += [str(r) for r in net.sinks]
    return " ".join(parts)


def _macro_lines(macro: Macro, indent: str) -> list[str]:
    lines = [f"{indent}macro {macro.id} {macro.module} {{"]
    lines += [
        f"{indent}{INDENT}{line}"
        for line in layout_lines(macro.layout, None, header=False)
    ]
    lines.append(f"{indent}}}")
    return lines


def _module_lines(module: Module, pack: str, indent: str) -> list[str]:
    lines = [f"{indent}module {module.id} {{"]
    for port in module.ports:
        lines.append(
            f"{indent}{INDENT}port {port.id} {port.direction} {port.carrier} = {port.inner}"
        )
    lines += scope_lines(module.body, pack, indent + INDENT)
    lines.append(f"{indent}}}")
    return lines


def scope_lines(netlist: Netlist, pack: str, indent: str = "") -> list[str]:
    lines = (
        [indent + " ".join(["attrs", *_attrs(netlist.attrs)])] if netlist.attrs else []
    )
    lines += _lib_lines(netlist, indent)
    for mid in sorted(netlist.modules):
        lines += _module_lines(netlist.modules[mid], pack, indent)
    for mid in sorted(netlist.macros):
        lines += _macro_lines(netlist.macros[mid], indent)
    for cid in sorted(netlist.cells):
        cell = netlist.cells[cid]
        lines.append(indent + _cell_line(cell, pack))
        lines += [indent + line for line in _pin_lines(netlist, cell)]
    for nid in sorted(netlist.nets):
        lines.append(indent + _net_line(netlist.nets[nid]))
    for gid in sorted(netlist.groups):
        group = netlist.groups[gid]
        parts = ["group", group.id]
        if group.kind:
            parts.append(_opt("kind", group.kind))
        parts += _attrs(group.attrs)
        lines.append(indent + " ".join([*parts, ":", *group.members]))
    return lines


# ------------------------------------------------------------------ layout


def _endpoint(cell_xy: tuple[int, int], pins: dict[tuple[int, int], PinRef]) -> str:
    ref = pins.get(cell_xy)
    return str(ref) if ref is not None else f"@{cell_xy[0]},{cell_xy[1]}"


def _wire_line(wire: Wire, layout: Layout, netlist: Netlist | None) -> str:
    net = netlist.nets.get(wire.net) if netlist is not None else None
    parts = ["wire", wire.net]
    attach: dict[tuple[int, int], PinRef] = {}
    if net is None or netlist is None:
        first = wire.segments[0] if wire.segments else None
        if first is not None:
            parts += [_opt("carrier", first.carrier), _opt("layer", first.layer)]
    else:
        for ref in net.sources + net.sinks:
            xy = layout.attach(netlist, ref)
            if xy is not None and xy not in attach:
                attach[xy] = ref
    if wire.units:
        parts.append(_opt("units", list(wire.units)))
    segments = []
    for segment in wire.segments:
        cells = segment.cells
        if net is None or netlist is None:
            body = (
                "cells " + " ".join(f"{x},{y}" for x, y in cells[:-1])
                if len(cells) > 1
                else "cells " + f"{cells[0][0]},{cells[0][1]}"
            )
            segments.append(f"{body} -> @{cells[-1][0]},{cells[-1][1]}")
        else:
            start = _endpoint(cells[0], attach)
            moves = moves_from_cells(cells)
            end = _endpoint(cells[-1], attach)
            segments.append(" ".join([start, *moves, "->", end]))
    return " ".join([*parts, ":", " ; ".join(segments)])


def layout_lines(
    layout: Layout, netlist: Netlist | None, header: bool = True
) -> list[str]:
    flat_netlist = netlist.flatten() if netlist is not None else None
    head = " ".join(p for p in ["layout", layout.problem, *_attrs(layout.attrs)] if p)
    lines = [head] if header else []
    for cid in sorted(layout.placements):
        p = layout.placements[cid]
        lines.append(f"place {p.cell} {p.x},{p.y} r{p.rot}")
    for cid in sorted(layout.instances):
        p = layout.instances[cid]
        lines.append(f"instance {p.cell} {p.x},{p.y} r{p.rot}")
    for nid in sorted(layout.wires):
        lines.append(_wire_line(layout.wires[nid], layout, flat_netlist))
    for uid in sorted(layout.units):
        unit = layout.units[uid]
        parts = [
            "unit",
            unit.id,
            unit.footprint,
            f"{unit.x},{unit.y}",
            f"r{unit.rot}",
            _opt("owner", unit.owner),
        ]
        if unit.kind and unit.kind != unit.footprint:
            parts.append(_opt("kind", unit.kind))
        parts += _attrs(unit.attrs)
        lines.append(" ".join(parts))
    for reservation in layout.reservations:
        parts = ["reserve", reservation.tag, reservation.layer]
        if reservation.carrier:
            parts.append(_opt("carrier", reservation.carrier))
        rects = Region.of("r", frozenset(reservation.cells)).rects
        lines.append(
            " ".join([*parts, ":", *[f"rect {r.x},{r.y} {r.w}x{r.h}" for r in rects]])
        )
    return lines


# -------------------------------------------------------------- assessment


def assessment_lines(assessment: Assessment) -> list[str]:
    lines = [f"assessment {assessment.layout}".rstrip()]
    for name in sorted(assessment.metrics):
        lines.append(f"metric {name} {write_value(assessment.metrics[name])}")
    for finding in assessment.findings:
        parts = [
            "finding",
            finding.rule,
            finding.severity,
            finding.subject,
            quote(finding.message),
            *_attrs(finding.attrs),
        ]
        lines.append(" ".join(parts))
    lines.append(f"complete {'true' if assessment.complete else 'false'}")
    lines.append(f"valid {'true' if assessment.valid else 'false'}")
    return lines


# ----------------------------------------------------------------- entry


def write(level: Level, context: Any = None) -> str:
    """The canonical text of ``level``; a layout gets pin endpoints when ``context`` holds the netlist."""
    context_json = None
    if isinstance(context, Problem | Netlist):
        context_json = context.to_json()
    elif context is not None and getattr(context, "problem", None) is not None:
        context_json = context.problem.to_json()
    elif context is not None and getattr(context, "netlist", None) is not None:
        context_json = context.netlist.to_json()
    native = rust_write_kl(level.to_json(), context_json)
    if native is not None:
        return native
    lines = ["kl 1"]
    if isinstance(level, Problem):
        if level.physics:
            lines.append(f"physics {level.physics}")
        lines += fabric_lines(level.fabric)
        for key in sorted(level.params):
            lines.append(f"param {key}={write_loose_value(level.params[key])}")
        lines += scope_lines(
            level.netlist,
            level.physics.split("@")[0] if level.physics else level.netlist.pack,
        )
    elif isinstance(level, Netlist):
        if level.pack:
            lines.append(f"physics {level.pack}")
        lines += scope_lines(level, level.pack)
    elif isinstance(level, Layout):
        netlist = getattr(context, "netlist", None) if context is not None else None
        if isinstance(context, Problem):
            netlist = context.netlist
        lines += layout_lines(level, netlist)
    elif isinstance(level, Assessment):
        lines += assessment_lines(level)
    else:
        raise TypeError(f"no text form for {type(level).__name__}")
    return "\n".join(lines) + "\n"

File: src/kohakulayout-rs/src/text/writer.rs
//! Levels to canonical text: sorted ids, fixed option order, moves for wires when the netlist is known.
//! A port of Python's `writer.py`, byte for byte.

use std::collections::BTreeMap;

use serde_json::Value;

use super::assemble::attach;
use super::moves::moves_from_cells;
use super::values::{quote, write_loose_value, write_metric, write_value};
use crate::ir::geometry::{rects_from_cells, XY};
use crate::ir::model::{
    Assessment, Attrs, Cell, Fabric, Layout, Macro, Module, Net, Netlist, Pin, PinRef, Problem,
    Region, Wire,
};
use crate::ir::rate::Rate;

const INDENT: &str = "    ";

fn opt(key: &str, value: &Value) -> String {
    format!("{key}={}", write_value(value))
}

fn opt_str(key: &str, value: &str) -> String {
    opt(key, &Value::String(value.to_string()))
}

fn opt_list(key: &str, values: &[String]) -> String {
    opt(key, &Value::Array(values.iter().map(|v| Value::String(v.clone())).collect()))
}

fn attrs(attrs: &Attrs) -> Vec<String> {
    let mut out = Vec::new();
    for (ns, values) in attrs {
        for (key, value) in values {
            out.push(format!("+{ns}.{key}={}", write_loose_value(value)));
        }
    }
    out
}

fn rate(value: &Rate) -> String {
    value.text()
}

// ------------------------------------------------------------------ fabric

fn region_line(region: &Region, fabric: &Fabric) -> String {
    let every = (fabric.width * fabric.height) as usize;
    let body = if region.cells().len() == every {
        "all".to_string()
    } else {
        region
            .rects
            .iter()
            .map(|r| format!("rect {},{} {}x{}", r.x, r.y, r.w, r.h))
            .collect::<Vec<_>>()
            .join(" ")
    };
    let mut parts = vec!["region".to_string(), region.id.clone(), body];
    parts.extend(attrs(&region.attrs));
    parts.join(" ")
}

pub fn fabric_lines(fabric: &Fabric) -> Vec<String> {
    let mut parts = vec![
        "fabric".to_string(),
        format!("{}x{}", fabric.width, fabric.height),
        opt_list("layers", &fabric.layers),
    ];
    if !fabric.entries.is_empty() {
        parts.push(opt_list("entries", &fabric.entries));
    }
    parts.extend(attrs(&fabric.attrs));
    let mut lines = vec![parts.join(" ")];
    for region in fabric.regions.values() {
        lines.push(region_line(region, fabric));
    }
    for carrier in fabric.carriers.values() {
        let mut parts = vec![
            "carrier".to_string(),
            carrier.id.clone(),
            carrier.layer.clone(),
        ];
        if let Some(capacity) = &carrier.capacity {
            parts.push(rate(capacity));
        }
        parts.extend(attrs(&carrier.attrs));
        lines.push(parts.join(" "));
    }
    lines
}

// ----------------------------------------------------------------- netlist

fn lib_lines(netlist: &Netlist, indent: &str) -> Vec<String> {
    let mut lines = Vec::new();
    for fp in netlist.library.values() {
        let mut parts = vec![
            "lib".to_string(),
            fp.id.clone(),
            format!("{}x{}", fp.width, fp.height),
        ];
        if fp.layer != "ground" {
            parts.push(opt_str("layer", &fp.layer));
        }
        if !fp.occludes.is_empty() {
            parts.push(opt_list("occludes", &fp.occludes));
        }
        if fp.rotations != [0, 90, 180, 270] {
            parts.push(opt(
                "rot",
                &Value::Array(
                    fp.rotations
                        .iter()
                        .map(|r| Value::Number((*r).into()))
                        .collect(),
                ),
            ));
        }
        parts.extend(attrs(&fp.attrs));
        let ports: Vec<String> = fp
            .ports
            .iter()
            .map(|p| {
                let mut line = vec![format!(
                    "{} {} {} {}{}",
                    p.id, p.direction, p.carrier, p.side, p.offset
                )];
                line.extend(attrs(&p.attrs));
                line.join(" ")
            })
            .collect();
        if ports.len() <= 4 {
            if ports.is_empty() {
                lines.push(format!("{indent}{} {{ }}", parts.join(" ")));
            } else {
                lines.push(format!("{indent}{} {{ {} }}", parts.join(" "), ports.join(" ; ")));
            }
        } else {
            lines.push(format!("{indent}{} {{", parts.join(" ")));
            for p in ports {
                lines.push(format!("{indent}{INDENT}{p}"));
            }
            lines.push(format!("{indent}}}"));
        }
    }
    lines
}

fn cell_line(cell: &Cell, pack: &str) -> String {
    let reference = cell
        .footprint
        .clone()
        .or_else(|| cell.module.clone())
        .or_else(|| cell.macro_.clone())
        .unwrap_or_default();
    let mut parts = vec!["cell".to_string(), cell.id.clone(), reference.clone()];
    if !cell.kind.is_empty() && cell.kind != reference {
        parts.push(opt_str("kind", &cell.kind));
    }
    if !cell.label.is_empty() {
        parts.push(opt_str("label", &cell.label));
    }
    let constraint = &cell.constraint;
    if constraint.kind != "free" || !constraint.attrs.is_empty() {
        parts.push(opt_str("at", &constraint.kind));
        let own = if pack.is_empty() { "pack" } else { pack };
        for (ns, values) in &constraint.attrs {
            for (key, value) in values {
                let name = if ns == own {
                    key.clone()
                } else {
                    format!("at.{ns}.{key}")
                };
                parts.push(opt(&name, value));
            }
        }
    }
    if let Some(group) = &cell.group {
        parts.push(opt_str("group", group));
    }
    if !cell.needs.is_empty() {
        parts.push(opt_list("needs", &cell.needs));
    }
    parts.extend(attrs(&cell.attrs));
    parts.join(" ")
}

fn default_pins(netlist: &Netlist, cell: &Cell) -> Vec<Pin> {
    let mut view = netlist.clone();
    let mut bare = cell.clone();
    bare.pins.clear();
    view.cells.insert(cell.id.clone(), bare);
    view.pins_of(&cell.id)
}

fn pin_lines(netlist: &Netlist, cell: &Cell) -> Vec<String> {
    if cell.pins.is_empty() || cell.pins == default_pins(netlist, cell) {
        return Vec::new();
    }
    cell.pins
        .iter()
        .map(|p| {
            format!(
                "pin {}.{} {} {} {}",
                cell.id,
                p.id,
                p.direction,
                p.carrier,
                opt_str("ports", &p.ports.join("|"))
            )
        })
        .collect()
}

fn net_line(net: &Net) -> String {
    let mut parts = vec!["net".to_string(), net.id.clone(), net.carrier.clone()];
    if !net.rate.is_zero() {
        parts.push(rate(&net.rate));
    }
    if !net.kind.is_empty() {
        parts.push(opt_str("kind", &net.kind));
    }
    if !net.label.is_empty() {
        parts.push(opt_str("label", &net.label));
    }
    if let Some(outside) = &net.outside {
        parts.push(opt_str("outside", outside));
    }
    parts.extend(attrs(&net.attrs));
    parts.push(":".to_string());
    parts.extend(net.sources.iter().map(PinRef::text));
    parts.push("->".to_string());
    parts.extend(net.sinks.iter().map(PinRef::text));
    parts.join(" ")
}

fn macro_lines(macro_: &Macro, indent: &str) -> Vec<String> {
    let mut lines = vec![format!("{indent}macro {} {} {{", macro_.id, macro_.module)];
    for line in layout_lines(&macro_.layout, None, false) {
        lines.push(format!("{indent}{INDENT}{line}"));
    }
    lines.push(format!("{indent}}}"));
    lines
}

fn module_lines(module: &Module, pack: &str, indent: &str) -> Vec<String> {
    let mut lines = vec![format!("{indent}module {} {{", module.id)];
    for port in &module.ports {
        lines.push(format!(
            "{indent}{INDENT}port {} {} {} = {}",
            port.id,
            port.direction,
            port.carrier,
            port.inner.text()
        ));
    }
    lines.extend(scope_lines(&module.body, pack, &format!("{indent}{INDENT}")));
    lines.push(format!("{indent}}}"));
    lines
}

pub fn scope_lines(netlist: &Netlist, pack: &str, indent: &str) -> Vec<String> {
    let mut lines = Vec::new();
    if !netlist.attrs.is_empty() {
        let mut parts = vec!["attrs".to_string()];
        parts.extend(attrs(&netlist.attrs));
        lines.push(format!("{indent}{}", parts.join(" ")));
    }
    lines.extend(lib_lines(netlist, indent));
    for module in netlist.modules.values() {
        lines.extend(module_lines(module, pack, indent));
    }
    for macro_ in netlist.macros.values() {
        lines.extend(macro_lines(macro_, indent));
    }
    for cell in netlist.cells.values() {
        lines.push(format!("{indent}{}", cell_line(cell, pack)));
        for line in pin_lines(netlist, cell) {
            lines.push(format!("{indent}{line}"));
        }
    }
    for net in netlist.nets.values() {
        lines.push(format!("{indent}{}", net_line(net)));
    }
    for group in netlist.groups.values() {
        let mut parts = vec!["group".to_string(), group.id.clone()];
        if !group.kind.is_empty() {
            parts.push(opt_str("kind", &group.kind));
        }
        parts.extend(attrs(&group.attrs));
        parts.push(":".to_string());
        parts.extend(group.members.iter().cloned());
        lines.push(format!("{indent}{}", parts.join(" ")));
    }
    lines
}

// ------------------------------------------------------------------ layout

fn endpoint(cell_xy: XY, pins: &BTreeMap<XY, PinRef>) -> String {
    match pins.get(&cell_xy) {
        Some(r) => r.text(),
        None => format!("@{},{}", cell_xy.0, cell_xy.1),
    }
}

fn wire_line(wire: &Wire, layout: &Layout, netlist: Option<&Netlist>) -> String {
    let net = netlist.and_then(|n| n.nets.get(&wire.net));
    let mut parts = vec!["wire".to_string(), wire.net.clone()];
    let mut pins: BTreeMap<XY, PinRef> = BTreeMap::new();
    match (net, netlist) {
        (Some(net), Some(netlist)) => {
            for r in net.sources.iter().chain(net.sinks.iter()) {
                if let Some(xy) = attach(layout, netlist, r) {
                    pins.entry(xy).or_insert_with(|| r.clone());
                }
            }
        }
        _ => {
            if let Some(first) = wire.segments.first() {
                parts.push(opt_str("carrier", &first.carrier));
                parts.push(opt_str("layer", &first.layer));
            }
        }
    }
    if !wire.units.is_empty() {
        parts.push(opt_list("units", &wire.units));
    }
    let mut segments = Vec::new();
    for segment in &wire.segments {
        let cells = &segment.cells;
        if net.is_none() || netlist.is_none() {
            let body = if cells.len() > 1 {
                format!(
                    "cells {}",
                    cells[..cells.len() - 1]
                        .iter()
                        .map(|(x, y)| format!("{x},{y}"))
                        .collect::<Vec<_>>()
                        .join(" ")
                )
            } else {
                format!("cells {},{}", cells[0].0, cells[0].1)
            };
            let last = cells.last().unwrap();
            segments.push(format!("{body} -> @{},{}", last.0, last.1));
        } else {
            let mut line = vec![endpoint(cells[0], &pins)];
            line.extend(moves_from_cells(cells));
            line.push("->".to_string());
            line.push(endpoint(*cells.last().unwrap(), &pins));
            segments.push(line.join(" "));
        }
    }
    parts.push(":".to_string());
    parts.push(segments.join(" ; "));
    parts.join(" ")
}

pub fn layout_lines(layout: &Layout, netlist: Option<&Netlist>, header: bool) -> Vec<String> {
    let flat = netlist.map(|n| n.flatten());
    let mut lines = Vec::new();
    if header {
        let mut head = vec!["layout".to_string()];
        if !layout.problem.is_empty() {
            head.push(layout.problem.clone());
        }
        head.extend(attrs(&layout.attrs));
        lines.push(head.join(" "));
    }
    for p in layout.placements.values() {
        lines.push(format!("place {} {},{} r{}", p.cell, p.x, p.y, p.rot));
    }
    for p in layout.instances.values() {
        lines.push(format!("instance {} {},{} r{}", p.cell, p.x, p.y, p.rot));
    }
    for wire in layout.wires.values() {
        lines.push(wire_line(wire, layout, flat.as_ref()));
    }
    for unit in layout.units.values() {
        let mut parts = vec![
            "unit".to_string(),
            unit.id.clone(),
            unit.footprint.clone(),
            format!("{},{}", unit.x, unit.y),
            format!("r{}", unit.rot),
            opt_str("owner", &unit.owner),
        ];
        if !unit.kind.is_empty() && unit.kind != unit.footprint {
            parts.push(opt_str("kind", &unit.kind));
        }
        parts.extend(attrs(&unit.attrs));
        lines.push(parts.join(" "));
    }
    for reservation in &layout.reservations {
        let mut parts = vec![
            "reserve".to_string(),
            reservation.tag.clone(),
            reservation.layer.clone(),
        ];
        if let Some(carrier) = &reservation.carrier {
            if !carrier.is_empty() {
                parts.push(opt_str("carrier", carrier));
            }
        }
        let cells: std::collections::BTreeSet<XY> = reservation.cells.iter().cloned().collect();
        parts.push(":".to_string());
        parts.extend(
            rects_from_cells(&cells)
                .iter()
                .map(|r| format!("rect {},{} {}x{}", r.x, r.y, r.w, r.h)),
        );
        lines.push(parts.join(" "));
    }
    lines
}

// -------------------------------------------------------------- assessment

pub fn assessment_lines(assessment: &Assessment) -> Vec<String> {
    let mut lines = vec![format!("assessment {}", assessment.layout)
        .trim_end()
        .to_string()];
    for (name, value) in &assessment.metrics {
        lines.push(format!("metric {name} {}", write_metric(value)));
    }
    for finding in &assessment.findings {
        let mut parts = vec![
            "finding".to_string(),
            finding.rule.clone(),
            finding.severity.clone(),
            finding.subject.clone(),
            quote(&finding.message),
        ];
        parts.extend(attrs(&finding.attrs));
        lines.push(parts.join(" "));
    }
    lines.push(format!("complete {}", if assessment.complete { "true" } else { "false" }));
    lines.push(format!("valid {}", if assessment.valid { "true" } else { "false" }));
    lines
}

// ----------------------------------------------------------------- entry

pub fn write_problem(problem: &Problem) -> String {
    let mut lines = vec!["kl 1".to_string()];
    if !problem.physics.is_empty() {
        lines.push(format!("physics {}", problem.physics));
    }
    lines.extend(fabric_lines(&problem.fabric));
    for (key, value) in &problem.params {
        lines.push(format!("param {key}={}", write_loose_value(value)));
    }
    let pack = if problem.physics.is_empty() {
        problem.netlist.pack.clone()
    } else {
        problem.physics.split('@').next().unwrap_or("").to_string()
    };
    lines.extend(scope_lines(&problem.netlist, &pack, ""));
    lines.join("\n") + "\n"
}

pub fn write_netlist(netlist: &Netlist) -> String {
    let mut lines = vec!["kl 1".to_string()];
    if !netlist.pack.is_empty() {
        lines.push(format!("physics {}", netlist.pack));
    }
    lines.extend(scope_lines(netlist, &netlist.pack, ""));
    lines.join("\n") + "\n"
}

pub fn write_layout(layout: &Layout, netlist: Option<&Netlist>) -> String {
    let mut lines = vec!["kl 1".to_string()];
    lines.extend(layout_lines(layout, netlist, true));
    lines.join("\n") + "\n"
}

pub fn write_assessment(assessment: &Assessment) -> String {
    let mut lines = vec!["kl 1".to_string()];
    lines.extend(assessment_lines(assessment));
    lines.join("\n") + "\n"
}
